tools/src/bin/fap-lld/fastrel.rs

```diff
@@ -0,0 +1,156 @@
+use std::{
+    collections::HashMap,
+    fs,
+    io::{self, Write},
+    path::Path,
+    process::Command,
+};
+
+use elf::{
+    abi::SHT_REL, endian::LittleEndian, relocation::RelIterator, string_table::StringTable,
+    symbol::SymbolTable, ElfBytes,
+};
+use md5::{Digest, Md5};
+use tempfile::tempdir;
+
+use super::Error;
+
+pub(crate) fn postprocess_fap(output_fap: &Path, objcopy: &Path) -> Result<(), Error> {
+    // Parse the FAP as an ELF binary.
+    let fap_data = fs::read(output_fap)?;
+    let fap = ElfBytes::<LittleEndian>::minimal_parse(&fap_data)?;
+
+    // Get the section header table alongside its string table.
+    let (shdrs_opt, strtab_opt) = fap.section_headers_with_strtab()?;
+    let (shdrs, strtab) = shdrs_opt.zip(strtab_opt).ok_or(Error::NoSectionHeaders)?;
+
+    // Collect the sections with relocations.
+    let rel_sections = shdrs
+        .iter()
+        .filter(|shdr| shdr.sh_type == SHT_REL)
+        .map(|shdr| -> Result<_, Error> {
+            let name = strtab.get(shdr.sh_name as usize)?;
+            let section = fap.section_data_as_rels(&shdr)?;
+            Ok((name, section))
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Convert the relocations into `.fast.rel` sections.
+    let (symtab, strtab) = fap.symbol_table()?.ok_or(Error::NoSymbolTable)?;
+    let fastrel_sections = rel_sections
+        .into_iter()
+        .map(|(section_name, section)| FastRelSection::new(section_name, section, &symtab, &strtab))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Write the `.fast.rel` sections into the binary.
+    let temp_dir = tempdir()?;
+    for section in fastrel_sections {
+        let fastrel_section_name = hex::encode(Md5::digest(&section.name));
+        let data_path = temp_dir
+            .path()
+            .join(fastrel_section_name)
+            .with_extension("bin");
+        let mut data = fs::OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(&data_path)?;
+        section.write(&mut data)?;
+        data.flush()?;
+
+        let res = Command::new(objcopy)
+            .arg("--add-section")
+            .arg(format!("{}={}", section.name, data_path.display()))
+            .arg(output_fap)
+            .status()?;
+        if !res.success() {
+            return Err(Error::ObjcopyFailed);
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct FastRel<'data> {
+    section_index: u16,
+    section_value: u64,
+    r_type: u32,
+    name: &'data str,
+}
+
+impl<'data> FastRel<'data> {
+    fn gnu_sym_hash(&self) -> u32 {
+        let mut h = 0x1505;
+        for c in self.name.as_bytes() {
+            h = (h << 5) + h + u32::from(*c);
+        }
+        h
```

Comment:
This is a nit-picky personal style thing: I find it helpful to have a blank line before the return value.

Visually it makes it a bit more distinct, which I find helpful when scanning a function. Thoughts?

Reply:
I usually do that as well, but this is such a small function that it didn't seem worthwhile. If I were altering this, I'd probably rewrite it to use `Iterator::fold` instead.